modules/epf/samples/epf_demo.cpp

```diff
@@ -0,0 +1,99 @@
+//
+// Copyright (C) 2000-2008, Intel Corporation, all rights reserved.
+// Copyright (C) 2009, Willow Garage Inc., all rights reserved.
+// Third party copyrights are property of their respective owners.
+//
+// Redistribution and use in source and binary forms, with or without modification,
+// are permitted provided that the following conditions are met:
+//
+//   * Redistribution's of source code must retain the above copyright notice,
+//     this list of conditions and the following disclaimer.
+//
+//   * Redistribution's in binary form must reproduce the above copyright notice,
+//     this list of conditions and the following disclaimer in the documentation
+//     and/or other materials provided with the distribution.
+//
+//   * The name of the copyright holders may not be used to endorse or promote products
+//     derived from this software without specific prior written permission.
+//
+// This software is provided by the copyright holders and contributors "as is" and
+// any express or implied warranties, including, but not limited to, the implied
+// warranties of merchantability and fitness for a particular purpose are disclaimed.
+// In no event shall the Intel Corporation or contributors be liable for any direct,
+// indirect, incidental, special, exemplary, or consequential damages
+// (including, but not limited to, procurement of substitute goods or services;
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+//################################################################################
+//
+//                    Created by Simon Reich
+//
+//################################################################################
+
+#include <iostream>
+#include <string>
+#include <opencv2/opencv.hpp>
+#include <opencv2/epf.hpp>
+
+using namespace cv;
+
+int showWindow(cv::Mat image, std::string title);
+
+int showWindow(cv::Mat image, std::string title)
+{
+	namedWindow(title);
+	imshow(title, image);
+
+	waitKey(0);
+
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	// Help text
+	if (argc != 2)
+	{
+		std::cout << "usage: " << argv[0] << " image.jpg" << std::endl;
+		std::cout << "image.jpg contains image, which will be smoothed."
+		          << std::endl;
+		return -1;
+	}
+
+	// Load image from first parameter
+	std::string filename = argv[1];
+	Mat image = imread(filename, 1), res;
+
+	if (!image.data)
+	{
+		std::cerr << "No image data at " << argv[2] << std::endl;
+		throw;
+	}
+
+	// Before filtering
+	showWindow(image, "Original image");
+
+	// Initialize filter. Kernel size 5x5, threshold 20
+	Ptr<epf::edgepreservingFilter> filter =
+	    epf::edgepreservingFilter::create(&image, &res, 5, 20);
```

Comment:
Looks like using "filter" object is over-complicated for filtering case. `Ptr` is not necessary.
It is better to use just a function, like OpenCV's [`cv::blur`](https://docs.opencv.org/3.4.2/d4/d86/group__imgproc__filter.html#ga8c45db9afe636703801b0b2e440fce37).

Please use OpenCV's InputArray / OutputArray types for parameters (they are used for Python/Java bindings generation at least)

Reply:
I did not know how to implement this properly. It is now a part of ximgproc.

InputArray is used.

No pointers are used.